core/net/mac/tsch/sixtop/sixp.h

```diff
@@ -0,0 +1,297 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/**
+ * \addtogroup sixtop
+ * @{
+ */
+/**
+ * \file
+ *         6top Protocol (6P) APIs
+ * \author
+ *         Yasuyuki Tanaka <yasuyuki.tanaka@inf.ethz.ch>
+ */
+
+#ifndef _SIXP_H_
+#define _SIXP_H_
+
+#include "net/linkaddr.h"
+#include "sys/clock.h"
+
+#define SIXP_SUBIE_ID 0x00
+#define SIXP_VERSION  0x01
+
+/**
+ * \brief 6P Message Types
+ */
+typedef enum  {
+  SIXP_TYPE_REQUEST      = 0b00, /**< 6P Request */
+  SIXP_TYPE_RESPONSE     = 0b01, /**< 6P Response */
+  SIXP_TYPE_CONFIRMATION = 0b10, /**< 6P Confirmation */
+  SIXP_TYPE_RESERVED     = 0b11, /**< Reserved */
+} sixp_type_t;
+
+/**
+ * \brief 6P Command Identifiers
+ */
+typedef enum {
+  SIXP_CMD_ADD       = 0x01, /**< CMD_ADD */
+  SIXP_CMD_DELETE    = 0x02, /**< CMD_DELETE */
+  SIXP_CMD_STATUS    = 0x03, /**< CMD_STATUS */
+  SIXP_CMD_LIST      = 0x04, /**< CMD_LIST */
+  SIXP_CMD_CLEAR     = 0x05, /**< CMD_CLEAR */
+} sixp_command_id_t;
+
+/**
+ * \brief 6P Return Codes
+ */
+typedef enum {
+  SIXP_RC_SUCCESS    = 0x06, /**< RC_SUCCESS */
+  SIXP_RC_ERR_VER    = 0x07, /**< RC_ERR_VER */
+  SIXP_RC_ERR_SFID   = 0x08, /**< RC_ERR_SFID */
+  SIXP_RC_ERR_GEN    = 0x09, /**< RC_ERR_GEN */
+  SIXP_RC_ERR_BUSY   = 0x0a, /**< RC_ERR_BUSY */
+  SIXP_RC_ERR_NORES  = 0x0b, /**< RC_ERR_NORES */
+  SIXP_RC_ERR_RESET  = 0x0c, /**< RC_ERR_RESET */
+  SIXP_RC_ERR        = 0x0d, /**< RC_ERR */
+} sixp_return_code_t;
+
+/**
+ * \brief 6P Codes integrating Command IDs and Return Codes
+ */
+typedef union {
+  sixp_command_id_t cmd; /**< 6P Command Identifier */
+  sixp_return_code_t rc; /**< 6P Return Code */
+  uint8_t value;         /**< 8-bit unsigned integer value */
+} sixp_code_t;
+
+/**
+ * \brief 6P Cell Options
+ */
+typedef enum {
+  SIXP_CELL_OPTION_TX     = 0x01, /**< TX Cell */
+  SIXP_CELL_OPTION_RX     = 0x02, /**< RX Cell */
+  SIXP_CELL_OPTION_SHARED = 0x04  /**< SHARED Cell */
+} sixp_cell_option_t;
+
+/**
+ * \brief 6P Schedule Generation Type
+ */
+typedef enum {
+  SIXP_GEN_TYPE_TX, /**< TX Cell */
+  SIXP_GEN_TYPE_RX  /**< RX Cell */
+} sixp_gen_type_t;
+
+
+/**
+ * \brief 6P Send Status, which represents sixp_send() result.
+ */
+typedef enum {
+  SIXP_SEND_STATUS_SUCCESS, /**< SUCCESS */
+  SIXP_SEND_STATUS_FAILURE  /**< FAILURE */
+} sixp_send_status_t;
+
+/**
+ * \brief 6P Add Request Format
+ */
+typedef struct {
+  uint16_t metadata;     /**< Metadata */
+  uint8_t cell_options;  /**< CellOptions */
+  uint8_t num_cells;     /**< NumCells */
+  uint32_t cell_list[0]; /**< CellList */
+} __attribute__((packed)) sixp_req_add_t;
```

Comment:
Pretty sure this had portability issues (across platforms and/or compilers).
The practice in uip6 is not to use struct for packet formatting. Fields are read/written one by one at a position guaranteed to be correct. Less efficient but 100% portable. For consistency with the rest of contiki I think 6p should follow the same practice.